File: src/core_modules/usage/query.rs
use super::sql::{self, Bound, Sql, literal};
use super::store::{self, Environment, MCP_AGENTS, SKILL_AGENTS, Session, identity};
use serde_json::{Map, Value, json};
use std::collections::{BTreeMap, HashMap};
use std::sync::OnceLock;

pub const SCHEMA_VERSION: i64 = 1;
pub const MAX_OBSERVED: usize = 64;
pub const WINDOW_DAYS: i64 = 160 * 7;
pub const UNAVAILABLE: &str = "usage store unavailable";
const MCP_EVENTS: &str = "(kind IN ('tool', 'resource', 'resource-list') OR (agent = 'claude' AND kind = 'command' AND name GLOB 'mcp__?*__?*'))";

fn definition_agent(agent: &str) -> Option<&'static str> {
    match agent {
        "claude" => Some("claude"),
        "codex" => Some("codex"),
        "opencode" => Some("opencode"),
        "github-copilot-cli" => Some("copilot"),
        _ => None,
    }
}

pub fn sanitize(name: &str) -> String {
    static NON_WORD: OnceLock<regex::Regex> = OnceLock::new();
    static RUNS: OnceLock<regex::Regex> = OnceLock::new();
    let non_word =
        NON_WORD.get_or_init(|| regex::Regex::new(r"[^a-zA-Z0-9_-]").expect("valid pattern"));
    let value = non_word.replace_all(name, "_").to_string();
    if !name.starts_with("claude.ai ") {
        return value;
    }
    let runs = RUNS.get_or_init(|| regex::Regex::new(r"_+").expect("valid pattern"));
    runs.replace_all(&value, "_").trim_matches('_').to_string()
}

fn field(item: &Value, key: &str) -> String {
    match item.get(key) {
        Some(Value::String(value)) => value.clone(),
        Some(Value::Null) | None => String::new(),
        Some(Value::Bool(false)) => String::new(),
        Some(other) => other.to_string(),
    }
}

pub fn skill_names(item: &Value) -> Vec<String> {
    let name = field(item, "name");
    let source = field(item, "source");
    let mut names = vec![name.clone()];
    if let Some(rest) = source.strip_prefix("plugin:") {
        let plugin = rest.split('@').next().unwrap_or("");
        let qualified = format!("{plugin}:{name}");
        if qualified != name {
            names.push(qualified);
        }
    }
    names
}

fn counts(agent: i64, user: i64, scheduled: i64, failed: i64) -> Map<String, Value> {
    let mut values = Map::new();
    values.insert("uses".to_string(), json!(agent + user));
    values.insert("usesAgent".to_string(), json!(agent));
    values.insert("usesUser".to_string(), json!(user));
    values.insert("usesScheduled".to_string(), json!(scheduled));
    values.insert("failed".to_string(), json!(failed));
    values
}

fn attach(item: &mut Value, values: &Map<String, Value>) {
    let Some(entries) = item.as_object_mut() else {
        return;
    };
    for (key, value) in values {
        entries.insert(key.clone(), value.clone());
    }
    let metrics = entries
        .entry("metrics".to_string())
        .or_insert_with(|| Value::Object(Map::new()));
    if !metrics.is_object() {
        *metrics = Value::Object(Map::new());
    }
    if let Some(metrics) = metrics.as_object_mut() {
        for (key, value) in values {
            metrics.insert(key.clone(), value.clone());
        }
    }
}

fn agent_list(agents: &[&str]) -> String {
    serde_json::to_string(agents).unwrap_or_else(|_| "[]".to_string())
}

fn sources(database: &Sql, agents: &[&str]) -> sql::Result<i64> {
    let statement = sql::bind(
        "SELECT count(*) FROM source WHERE agent IN (SELECT value FROM json_each(?))",
        &[Bound::Text(agent_list(agents))],
    )?;
    Ok(database
        .query_one(&statement)?
        .map_or(0, |row| row.integer(0)))
}

type Tallies = HashMap<String, [i64; 4]>;

fn skill_tallies(database: &Sql, where_clause: &str, parameters: &[Bound]) -> sql::Result<Tallies> {
    let statement = sql::bind(
        &format!(
            "SELECT kind, name, origin, count(*), sum(failed) FROM event WHERE kind IN ('skill', 'command') \
             {where_clause} GROUP BY kind, name, origin"
        ),
        parameters,
    )?;
    let mut tallies: Tallies = HashMap::new();
    for row in database.query(&statement)? {
        let (kind, name, origin, total, failed) = (
            row.text(0),
            row.text(1),
            row.text(2),
            row.integer(3),
            row.integer(4),
        );
        let tally = tallies.entry(name).or_insert([0; 4]);
        let slot = if kind == "skill" {
            0
        } else if origin == "user" {
            1
        } else {
            2
        };
        tally[slot] += total;
        if kind == "skill" {
            tally[3] += failed;
        }
    }
    Ok(tallies)
}

fn item_counts(item: &Value, tallies: &Tallies) -> Map<String, Value> {
    let mut total = [0i64; 4];
    let mut seen: Vec<String> = Vec::new();
    for name in skill_names(item) {
        if seen.contains(&name) {
            continue;
        }
        let tally = tallies.get(&name).copied().unwrap_or([0; 4]);
        for slot in 0..4 {
            total[slot] += tally[slot];
        }
        seen.push(name);
    }
    counts(total[0], total[1], total[2], total[3])
}

fn unavailable_extra() -> Map<String, Value> {
    let mut document = Map::new();
    document.insert("usageError".to_string(), Value::from(UNAVAILABLE));
    document
}

pub fn attach_skills(environment: &Environment, items: &mut [Value]) -> Map<String, Value> {
    match attached_skills(environment, items) {
        Ok(document) => document,
        Err(_) => unavailable_extra(),
    }
}

fn attached_skills(
    environment: &Environment,
    items: &mut [Value],
) -> sql::Result<Map<String, Value>> {
    let session = store::session(environment, true)?;
    let tallies = skill_tallies(&session.database, "", &[])?;
    for item in items.iter_mut() {
        let values = item_counts(item, &tallies);
        attach(item, &values);
    }
    let transcripts = sources(&session.database, &SKILL_AGENTS)?;
    let mut document = Map::new();
    document.insert("usageTranscripts".to_string(), json!(transcripts));
    document.insert("usageUnreadable".to_string(), json!(session.unreadable));
    document.insert("usageIngestPending".to_string(), json!(session.pending));
    Ok(document)
}

pub fn skill_counts(environment: &Environment, items: &[Value]) -> Value {
    match counted_skills(environment, items) {
        Ok(document) => document,
        Err(_) => json!({"ok": false, "schemaVersion": SCHEMA_VERSION, "error": UNAVAILABLE}),
    }
}

fn counted_skills(environment: &Environment, items: &[Value]) -> sql::Result<Value> {
    let session = store::session(environment, true)?;
    let tallies = skill_tallies(&session.database, "", &[])?;
    let mut counted = Map::new();
    for item in items {
        let id = field(item, "id");
        if id.is_empty() {
            continue;
        }
        counted.insert(id, Value::Object(item_counts(item, &tallies)));
    }
    let transcripts = sources(&session.database, &SKILL_AGENTS)?;
    Ok(json!({
        "ok": true,
        "schemaVersion": SCHEMA_VERSION,
        "counts": Value::Object(counted),
        "usageTranscripts": transcripts,
        "usageUnreadable": session.unreadable,
        "usageIngestPending": session.pending,
    }))
}

fn valid_day(day: &str) -> bool {
    day.len() == 10
        && day.as_bytes()[4] == b'-'
        && day.as_bytes()[7] == b'-'
        && day
            .bytes()
            .enumerate()
            .all(|(index, byte)| index == 4 || index == 7 || byte.is_ascii_digit())
}

pub fn skill_day(environment: &Environment, items: &[Value], day: &str) -> Value {
    if !valid_day(day) {
        return json!({"ok": false, "schemaVersion": SCHEMA_VERSION, "error": "day must be YYYY-MM-DD"});
    }
    match day_of_skills(environment, items, day) {
        Ok(document) => document,
        Err(_) => json!({"ok": false, "schemaVersion": SCHEMA_VERSION, "error": UNAVAILABLE}),
    }
}

fn day_of_skills(environment: &Environment, items: &[Value], day: &str) -> sql::Result<Value> {
    let session = store::session(environment, false)?;
    let tallies = skill_tallies(
        &session.database,
        "AND at >= CAST(strftime('%s', ?, 'utc') AS INTEGER) * 1000 \
         AND at < CAST(strftime('%s', ?, '+1 day', 'utc') AS INTEGER) * 1000",
        &[Bound::from(day), Bound::from(day)],
    )?;
    let mut used: Vec<(i64, String, Value)> = Vec::new();
    for item in items {
        let values = item_counts(item, &tallies);
        let uses = values.get("uses").and_then(Value::as_i64).unwrap_or(0);
        let scheduled = values
            .get("usesScheduled")
            .and_then(Value::as_i64)
            .unwrap_or(0);
        if uses <= 0 && scheduled <= 0 {
            continue;
        }
        let name = field(item, "name");
        let mut row = Map::new();
        row.insert("id".to_string(), Value::from(field(item, "id")));
        row.insert("name".to_string(), Value::from(name.clone()));
        for (key, value) in &values {
            row.insert(key.clone(), value.clone());
        }
        used.push((uses, name, Value::Object(row)));
    }
    used.sort_by(|left, right| right.0.cmp(&left.0).then_with(|| left.1.cmp(&right.1)));
    let rows: Vec<Value> = used.into_iter().map(|(_, _, row)| row).collect();
    Ok(json!({
        "ok": true,
        "schemaVersion": SCHEMA_VERSION,
        "day": day,
        "items": rows,
    }))
}

fn opencode_owner(name: &str, prefixes: &[String]) -> (String, String) {
    for prefix in prefixes {
        let head = format!("{prefix}_");
        if name.starts_with(&head) && name.len() > head.len() {
            return (prefix.clone(), name[head.len()..].to_string());
        }
    }
    (String::new(), name.to_string())
}

fn mcp_server(item: &Value) -> Option<(String, String)> {
    let name = field(item, "name");
    let agent = definition_agent(&field(item, "agent"))?;
    if agent == "opencode" {
        return Some((agent.to_string(), sanitize(&name)));
    }
    if agent != "claude" {
        return Some((agent.to_string(), name));
    }
    if field(item, "scope") != "plugin" {
        return Some((agent.to_string(), sanitize(&name)));
    }
    let plugin = match item.get("source") {
        Some(source @ Value::Object(_)) => field(source, "plugin"),
        _ => String::new(),
    };
    (!plugin.is_empty()).then(|| {
        (
            agent.to_string(),
            format!("plugin_{}_{}", sanitize(&plugin), sanitize(&name)),
        )
    })
}

#[derive(Default, Clone)]
struct Observation {
    agent: i64,
    user: i64,
    scheduled: i64,
    failed: i64,
    last: String,
}

pub fn attach_mcp(environment: &Environment, definitions: &mut [Value]) -> Map<String, Value> {
    match attached_mcp(environment, definitions) {
        Ok(document) => document,
        Err(_) => unavailable_extra(),
    }
}

fn attached_mcp(
    environment: &Environment,
    definitions: &mut [Value],
) -> sql::Result<Map<String, Value>> {
    let session = store::session(environment, true)?;
    let keys: Vec<Option<(String, String)>> = definitions.iter().map(mcp_server).collect();
    let mut prefixes: Vec<String> = keys
        .iter()
        .flatten()
        .filter(|(agent, _)| agent == "opencode")
        .map(|(_, server)| server.clone())
        .collect();
    prefixes.sort();
    prefixes.dedup();
    prefixes.sort_by(|left, right| right.len().cmp(&left.len()).then_with(|| left.cmp(right)));
    let mut servers: HashMap<(String, String), BTreeMap<(String, String), Observation>> =
        HashMap::new();
    {
        let statement = format!(
            "SELECT agent, server, kind, name, origin, count(*), sum(failed), \
             date(max(at) / 1000, 'unixepoch', 'localtime') FROM event WHERE {MCP_EVENTS} \
             GROUP BY agent, server, kind, name, origin"
        );
        for row in session.database.query(&statement)? {
            let (agent, mut server, mut kind, mut name, origin, total, failed, last) = (
                row.text(0),
                row.text(1),
                row.text(2),
                row.text(3),
                row.text(4),
                row.integer(5),
                row.integer(6),
                row.text(7),
            );
            if kind == "command" {
                let mut parts = name.splitn(3, "__");
                parts.next();
                server = parts.next().unwrap_or("").to_string();
                name = parts.next().unwrap_or("").to_string();
                kind = "prompt".to_string();
            }
            if agent == "claude" {
                server = sanitize(&server);
            } else if agent == "opencode" && server.is_empty() {
                let (owner, rest) = opencode_owner(&name, &prefixes);
                server = owner;
                name = rest;
            }
            let entry = servers
                .entry((agent, server))
                .or_default()
                .entry((kind.clone(), name))
                .or_default();
            if kind != "prompt" {
                entry.agent += total;
            } else if origin == "user" {
                entry.user += total;
            } else {
                entry.scheduled += total;
            }
            entry.failed += failed;
            if last > entry.last {
                entry.last = last;
            }
        }
    }
    let mut owners: HashMap<&(String, String), usize> = HashMap::new();
    for key in keys.iter().flatten() {
        *owners.entry(key).or_insert(0) += 1;
    }
    let mut ambiguous = 0i64;
    for (item, key) in definitions.iter_mut().zip(keys.iter()) {
        let mut observed = key
            .as_ref()
            .and_then(|key| servers.get(key))
            .cloned()
            .unwrap_or_default();
        if let Some(key) = key
            && owners.get(key).copied().unwrap_or(0) > 1
        {
            ambiguous += 1;
            if let Some(entries) = item.as_object_mut() {
                entries.insert("usageAmbiguous".to_string(), Value::Bool(true));
            }
            observed.clear();
        }
        let totals = observed.values().fold([0i64; 4], |mut total, entry| {
            total[0] += entry.agent;
            total[1] += entry.user;
            total[2] += entry.scheduled;
            total[3] += entry.failed;
            total
        });
        let values = counts(totals[0], totals[1], totals[2], totals[3]);
        attach(item, &values);
        let mut ranked: Vec<((String, String), Observation)> = observed.into_iter().collect();
        ranked.sort_by(|left, right| {
            let left_uses = -(left.1.agent + left.1.user);
            let right_uses = -(right.1.agent + right.1.user);
            left_uses
                .cmp(&right_uses)
                .then_with(|| left.0.1.cmp(&right.0.1))
                .then_with(|| left.0.0.cmp(&right.0.0))
        });
        let rows: Vec<Value> = ranked
            .into_iter()
            .take(MAX_OBSERVED)
            .map(|((kind, name), entry)| {
                json!({
                    "kind": kind,
                    "name": name,
                    "uses": entry.agent + entry.user,
                    "failed": entry.failed,
                    "lastUsed": entry.last,
                })
            })
            .collect();
        if let Some(entries) = item.as_object_mut() {
            entries.insert("observed".to_string(), Value::Array(rows));
        }
    }
    let transcripts = sources(&session.database, &MCP_AGENTS)?;
    let mut document = Map::new();
    document.insert("usageTranscripts".to_string(), json!(transcripts));
    document.insert("usageUnreadable".to_string(), json!(session.unreadable));
    document.insert("usageIngestPending".to_string(), json!(session.pending));
    document.insert("usageAmbiguous".to_string(), json!(ambiguous));
    Ok(document)
}

fn history(
    environment: &Environment,
    kind: &str,
    agents: &[&str],
    where_clause: &str,
    parameters: &[Bound],
    ingest: bool,
) -> Value {
    match tallied_history(environment, kind, agents, where_clause, parameters, ingest) {
        Ok(document) => document,
        Err(_) => {
            json!({"ok": false, "schemaVersion": SCHEMA_VERSION, "kind": kind, "error": UNAVAILABLE})
        }
    }
}

fn tallied_history(
    environment: &Environment,
    kind: &str,
    agents: &[&str],
    where_clause: &str,
    parameters: &[Bound],
    ingest: bool,
) -> sql::Result<Value> {
    let session: Session = store::session(environment, ingest)?;
    let coverage = sql::bind(
        "SELECT date(min(first_at) / 1000, 'unixepoch', 'localtime'), date('now', 'localtime') FROM coverage \
         WHERE agent IN (SELECT value FROM json_each(?))",
        &[Bound::Text(agent_list(agents))],
    )?;
    let (start, until): (Option<String>, Option<String>) = session
        .database
        .query_one(&coverage)?
        .map_or((None, None), |row| {
            (row.optional_text(0), row.optional_text(1))
        });
    let statement = sql::bind(
        &format!(
            "SELECT date(at / 1000, 'unixepoch', 'localtime') AS day, sum(kind <> 'command' OR origin = 'user'), \
             sum(kind <> 'command'), sum(kind = 'command' AND origin = 'user'), \
             sum(kind = 'command' AND origin = 'scheduled'), sum(failed) \
             FROM event WHERE {where_clause} AND day > date('now', 'localtime', '-{WINDOW_DAYS} days') \
             AND day <= date('now', 'localtime') GROUP BY day ORDER BY day"
        ),
        parameters,
    )?;
    let days: Vec<Value> = session
        .database
        .query(&statement)?
        .iter()
        .map(|row| {
            json!([
                row.text(0),
                row.integer(1),
                row.integer(2),
                row.integer(3),
                row.integer(4),
                row.integer(5),
            ])
        })
        .collect();
    Ok(json!({
        "ok": true,
        "schemaVersion": SCHEMA_VERSION,
        "kind": kind,
        "coverageStart": start,
        "until": until,
        "ingestPending": session.pending,
        "days": days,
    }))
}

pub fn skill_usage(environment: &Environment, items: &[Value], scoped: bool) -> Value {
    let mut names: Vec<String> = items.iter().flat_map(skill_names).collect();
    names.sort();
    names.dedup();
    let encoded = serde_json::to_string(&names).unwrap_or_else(|_| "[]".to_string());
    if scoped {
        return history(
            environment,
            "skill",
            &SKILL_AGENTS,
            "kind IN ('skill', 'command') AND name IN (SELECT value FROM json_each(?))",
            &[Bound::Text(encoded)],
            false,
        );
    }
    history(
        environment,
        "skill",
        &SKILL_AGENTS,
        "(kind = 'skill' OR (kind = 'command' AND name IN (SELECT value FROM json_each(?))))",
        &[Bound::Text(encoded)],
        true,
    )
}

pub fn mcp_usage(environment: &Environment) -> Value {
    history(environment, "mcp", &MCP_AGENTS, MCP_EVENTS, &[], true)
}

pub fn forget(environment: &Environment, before: Option<&str>) -> Value {
    match forgotten(environment, before) {
        Ok(document) => document,
        Err(_) => {
            json!({"ok": false, "schemaVersion": SCHEMA_VERSION, "removed": 0, "error": UNAVAILABLE})
        }
    }
}

fn forgotten(environment: &Environment, before: Option<&str>) -> sql::Result<Value> {
    let session = store::session(environment, false)?;
    let database = &session.database;
    let (cutoff, prelude, removal) = match before {
        Some(before) => {
            let statement = sql::bind(
                "SELECT CAST(strftime('%s', ?, 'utc') AS INTEGER) * 1000",
                &[Bound::from(before)],
            )?;
            let cutoff = database
                .query_one(&statement)?
                .and_then(|row| row.optional_integer(0))
                .ok_or_else(|| sql::Error::new("usage cutoff is not a day"))?;
            let value = literal(&Bound::Integer(cutoff));
            (
                cutoff,
                format!(
                    "DELETE FROM failure WHERE at < {value};\n\
                     UPDATE coverage SET first_at = max(first_at, {value});\n"
                ),
                format!("DELETE FROM event WHERE at < {value};\n"),
            )
        }
        None => {
            let cutoff = std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .map(|elapsed| elapsed.as_millis() as i64)
                .unwrap_or(0)
                + 1;
            let statement = sql::bind(
                "SELECT agent, call FROM event WHERE at >= ? UNION \
                 SELECT agent, call FROM failure WHERE at >= ?",
                &[Bound::Integer(cutoff), Bound::Integer(cutoff)],
            )?;
            let mut prelude = String::new();
            for row in database.query(&statement)? {
                prelude.push_str(&format!(
                    "INSERT OR IGNORE INTO forgotten VALUES ({});\n",
                    literal(&Bound::Blob(identity(&row.text(0), &row.text(1))))
                ));
            }
            prelude.push_str(
                "DELETE FROM failure;\nDELETE FROM coverage;\n\
                 UPDATE source SET project = NULL;\nDELETE FROM project;\n",
            );
            (cutoff, prelude, "DELETE FROM event;\n".to_string())
        }
    };
    let script = format!(
        "PRAGMA secure_delete = ON;\nBEGIN IMMEDIATE;\n{prelude}\
         INSERT INTO retention VALUES (1, {}) ON CONFLICT (id) DO UPDATE SET before = max(before, excluded.before);\n\
         {removal}SELECT changes();\nCOMMIT;\n",
        literal(&Bound::Integer(cutoff))
    );
    let removed = database.query_one(&script)?.map_or(0, |row| row.integer(0));
    Ok(json!({"ok": true, "schemaVersion": SCHEMA_VERSION, "removed": removed}))
}

File: python/agent_usage/query.py
from __future__ import annotations

import json
import re
import sqlite3
import time
from collections import Counter
from typing import Any

from .store import MCP_AGENTS, SKILL_AGENTS, identity, session

SCHEMA_VERSION = 1
MAX_OBSERVED = 64
WINDOW_DAYS = 160 * 7
UNAVAILABLE = "usage store unavailable"
MCP_EVENTS = "(kind IN ('tool', 'resource', 'resource-list') OR (agent = 'claude' AND kind = 'command' AND name GLOB 'mcp__?*__?*'))"
MCP_DEFINITION_AGENTS = {"claude": "claude", "codex": "codex", "opencode": "opencode", "github-copilot-cli": "copilot"}


def sanitize(name: str) -> str:
    value = re.sub(r"[^a-zA-Z0-9_-]", "_", name)
    return re.sub(r"_+", "_", value).strip("_") if name.startswith("claude.ai ") else value


def skill_names(item: dict[str, Any]) -> set[str]:
    name = str(item.get("name") or "")
    source = str(item.get("source") or "")
    return {name, source[len("plugin:"):].split("@", 1)[0] + ":" + name} if source.startswith("plugin:") else {name}


def mcp_server(item: dict[str, Any]) -> tuple[str, str] | None:
    name = str(item.get("name") or "")
    agent = MCP_DEFINITION_AGENTS.get(str(item.get("agent") or ""))
    if agent is None:
        return None
    if agent == "opencode":
        return agent, sanitize(name)
    if agent != "claude":
        return agent, name
    if item.get("scope") != "plugin":
        return agent, sanitize(name)
    plugin = str((item.get("source") or {}).get("plugin") or "")
    return (agent, f"plugin_{sanitize(plugin)}_{sanitize(name)}") if plugin else None


def counts(agent: int, user: int, scheduled: int, failed: int) -> dict[str, int]:
    return {"uses": agent + user, "usesAgent": agent, "usesUser": user, "usesScheduled": scheduled, "failed": failed}


def attach(item: dict[str, Any], values: dict[str, int]) -> None:
    item.update(values)
    item.setdefault("metrics", {}).update(values)


def sources(connection: sqlite3.Connection, agents: tuple[str, ...]) -> int:
    return connection.execute("SELECT count(*) FROM source WHERE agent IN (SELECT value FROM json_each(?))",
                              (json.dumps(agents),)).fetchone()[0]


def skill_tallies(connection: sqlite3.Connection, where: str = "", parameters: tuple[Any, ...] = ()) -> dict[str, list[int]]:
    tallies: dict[str, list[int]] = {}
    for kind, name, origin, total, failed in connection.execute(
            "SELECT kind, name, origin, count(*), sum(failed) FROM event WHERE kind IN ('skill', 'command') "
            f"{where} GROUP BY kind, name, origin", parameters):
        tally = tallies.setdefault(name, [0, 0, 0, 0])
        tally[0 if kind == "skill" else 1 if origin == "user" else 2] += total
        tally[3] += failed if kind == "skill" else 0
    return tallies


def item_counts(item: dict[str, Any], tallies: dict[str, list[int]]) -> dict[str, int]:
    return counts(*(sum(column) for column in zip(*(tallies.get(name, [0, 0, 0, 0]) for name in skill_names(item)))))


def attach_skills(items: list[dict[str, Any]]) -> dict[str, Any]:
    try:
        with session() as (connection, (pending, unreadable)):
            tallies = skill_tallies(connection)
            for item in items:
                attach(item, item_counts(item, tallies))
            return {"usageTranscripts": sources(connection, SKILL_AGENTS), "usageUnreadable": unreadable, "usageIngestPending": pending}
    except (OSError, sqlite3.Error):
        return {"usageError": UNAVAILABLE}


def skill_counts(items: list[dict[str, Any]]) -> dict[str, Any]:
    try:
        with session() as (connection, (pending, unreadable)):
            tallies = skill_tallies(connection)
            return {"ok": True, "schemaVersion": SCHEMA_VERSION,
                    "counts": {str(item.get("id") or ""): item_counts(item, tallies) for item in items if item.get("id")},
                    "usageTranscripts": sources(connection, SKILL_AGENTS), "usageUnreadable": unreadable, "usageIngestPending": pending}
    except (OSError, sqlite3.Error):
        return {"ok": False, "schemaVersion": SCHEMA_VERSION, "error": UNAVAILABLE}


def skill_day(items: list[dict[str, Any]], day: str) -> dict[str, Any]:
    if not re.fullmatch(r"\d{4}-\d{2}-\d{2}", day):
        return {"ok": False, "schemaVersion": SCHEMA_VERSION, "error": "day must be YYYY-MM-DD"}
    try:
        with session(ingest_history=False) as (connection, _):
            tallies = skill_tallies(connection,
                                    "AND at >= CAST(strftime('%s', ?, 'utc') AS INTEGER) * 1000 "
                                    "AND at < CAST(strftime('%s', ?, '+1 day', 'utc') AS INTEGER) * 1000", (day, day))
            used = []
            for item in items:
                values = item_counts(item, tallies)
                if values["uses"] > 0 or values["usesScheduled"] > 0:
                    used.append({"id": str(item.get("id") or ""), "name": str(item.get("name") or ""), **values})
            used.sort(key=lambda entry: (-entry["uses"], entry["name"]))
            return {"ok": True, "schemaVersion": SCHEMA_VERSION, "day": day, "items": used}
    except (OSError, sqlite3.Error):
        return {"ok": False, "schemaVersion": SCHEMA_VERSION, "error": UNAVAILABLE}


def opencode_owner(name: str, prefixes: list[str]) -> tuple[str, str]:
    for prefix in prefixes:
        if name.startswith(prefix + "_") and len(name) > len(prefix) + 1:
            return prefix, name[len(prefix) + 1:]
    return "", name


def attach_mcp(definitions: list[dict[str, Any]]) -> dict[str, Any]:
    try:
        with session() as (connection, (pending, unreadable)):
            keys = [mcp_server(item) for item in definitions]
            prefixes = sorted({server for key in keys if key and key[0] == "opencode" for server in [key[1]]}, key=len, reverse=True)
            servers: dict[tuple[str, str], dict[tuple[str, str], list[Any]]] = {}
            for agent, server, kind, name, origin, total, failed, last in connection.execute(
                    "SELECT agent, server, kind, name, origin, count(*), sum(failed), "
                    f"date(max(at) / 1000, 'unixepoch', 'localtime') FROM event WHERE {MCP_EVENTS} "
                    "GROUP BY agent, server, kind, name, origin"):
                if kind == "command":
                    _, server, name = name.split("__", 2)
                    kind = "prompt"
                if agent == "claude":
                    server = sanitize(server)
                elif agent == "opencode" and server == "":
                    server, name = opencode_owner(name, prefixes)
                entry = servers.setdefault((agent, server), {}).setdefault((kind, name), [0, 0, 0, 0, ""])
                entry[0 if kind != "prompt" else 1 if origin == "user" else 2] += total
                entry[3] += failed
                entry[4] = max(entry[4], last)
            owners = Counter(key for key in keys if key)
            ambiguous = 0
            for item, key in zip(definitions, keys):
                observed = servers.get(key, {}) if key else {}
                if key and owners[key] > 1:
                    ambiguous += 1
                    item["usageAmbiguous"] = True
                    observed = {}
                attach(item, counts(*(sum(entry[slot] for entry in observed.values()) for slot in range(4))))
                ranked = sorted(observed.items(), key=lambda pair: (-pair[1][0] - pair[1][1], pair[0][1], pair[0][0]))
                item["observed"] = [{"kind": kind, "name": name, "uses": entry[0] + entry[1], "failed": entry[3], "lastUsed": entry[4]}
                                    for (kind, name), entry in ranked[:MAX_OBSERVED]]
            return {"usageTranscripts": sources(connection, MCP_AGENTS), "usageUnreadable": unreadable,
                    "usageIngestPending": pending, "usageAmbiguous": ambiguous}
    except (OSError, sqlite3.Error):
        return {"usageError": UNAVAILABLE}


def history(kind: str, agents: tuple[str, ...], where: str, parameters: tuple[Any, ...] = (), *, ingest: bool = True) -> dict[str, Any]:
    try:
        with session(ingest_history=ingest) as (connection, (pending, _)):
            start, until = connection.execute(
                "SELECT date(min(first_at) / 1000, 'unixepoch', 'localtime'), date('now', 'localtime') FROM coverage "
                "WHERE agent IN (SELECT value FROM json_each(?))", (json.dumps(agents),)).fetchone()
            days = connection.execute(
                "SELECT date(at / 1000, 'unixepoch', 'localtime') AS day, sum(kind <> 'command' OR origin = 'user'), "
                "sum(kind <> 'command'), sum(kind = 'command' AND origin = 'user'), "
                "sum(kind = 'command' AND origin = 'scheduled'), sum(failed) "
                f"FROM event WHERE {where} AND day > date('now', 'localtime', '-{WINDOW_DAYS} days') "
                "AND day <= date('now', 'localtime') GROUP BY day ORDER BY day", parameters).fetchall()
            return {"ok": True, "schemaVersion": SCHEMA_VERSION, "kind": kind, "coverageStart": start, "until": until,
                    "ingestPending": pending, "days": [list(day) for day in days]}
    except (OSError, sqlite3.Error):
        return {"ok": False, "schemaVersion": SCHEMA_VERSION, "kind": kind, "error": UNAVAILABLE}


def skill_usage(items: list[dict[str, Any]], *, scoped: bool = False) -> dict[str, Any]:
    names = sorted({name for item in items for name in skill_names(item)})
    if scoped:
        return history("skill", SKILL_AGENTS,
                       "kind IN ('skill', 'command') AND name IN (SELECT value FROM json_each(?))", (json.dumps(names),), ingest=False)
    return history("skill", SKILL_AGENTS,
                   "(kind = 'skill' OR (kind = 'command' AND name IN (SELECT value FROM json_each(?))))", (json.dumps(names),))


def mcp_usage() -> dict[str, Any]:
    return history("mcp", MCP_AGENTS, MCP_EVENTS)


def forget(before: str | None) -> dict[str, Any]:
    try:
        with session(ingest_history=False) as (connection, _):
            connection.execute("PRAGMA secure_delete = ON")
            with connection:
                connection.execute("BEGIN IMMEDIATE")
                if before:
                    cutoff = connection.execute("SELECT CAST(strftime('%s', ?, 'utc') AS INTEGER) * 1000", (before,)).fetchone()[0]
                    removed = connection.execute("DELETE FROM event WHERE at < ?", (cutoff,)).rowcount
                    connection.execute("DELETE FROM failure WHERE at < ?", (cutoff,))
                    connection.execute("UPDATE coverage SET first_at = max(first_at, ?)", (cutoff,))
                else:
                    cutoff = time.time_ns() // 1_000_000 + 1
                    future = connection.execute("SELECT agent, call FROM event WHERE at >= ? UNION "
                                                "SELECT agent, call FROM failure WHERE at >= ?", (cutoff, cutoff)).fetchall()
                    connection.executemany("INSERT OR IGNORE INTO forgotten VALUES (?)", [(identity(*row),) for row in future])
                    removed = connection.execute("DELETE FROM event").rowcount
                    for statement in ("DELETE FROM failure", "DELETE FROM coverage", "UPDATE source SET project = NULL", "DELETE FROM project"):
                        connection.execute(statement)
                connection.execute("INSERT INTO retention VALUES (1, ?) ON CONFLICT (id) DO UPDATE "
                                   "SET before = max(before, excluded.before)", (cutoff,))
            return {"ok": True, "schemaVersion": SCHEMA_VERSION, "removed": removed}
    except (OSError, sqlite3.Error):
        return {"ok": False, "schemaVersion": SCHEMA_VERSION, "removed": 0, "error": UNAVAILABLE}
